File: homework4/Python/hsinyu_chang_task1.py
import sys
import os
os.environ["PYSPARK_SUBMIT_ARGS"] = (
    "--packages graphframes:graphframes:0.6.0-spark2.3-s_2.11")

from graphframes import *
from pyspark import SparkConf, SparkContext, StorageLevel
from pyspark.sql import SQLContext, SparkSession

from itertools import combinations
from operator import add
import time

#spark configuration
conf = SparkConf()
conf.setAppName('hw4')
conf.setMaster('local')
conf.set("spark.executor.memory", "4g")
conf.set("spark.driver.memory", "4g")
conf.set("spark.sql.shuffle.partitions", "4")

sc = SparkContext.getOrCreate(conf)
sqlContext = SQLContext(sc)

# arguments
threshold = int(sys.argv[1])
inputFilePath = sys.argv[2]
outputFilePath = sys.argv[3]

start = time.time()
inputFile = sc.textFile(inputFilePath)
header = inputFile.first()

#user_id,  business_id
inputRDD = inputFile.filter(lambda row: row!=header).map(lambda row: row.split(',')).map(lambda row: (row[0], row[1])).persist(StorageLevel.DISK_ONLY)

edgesRDD = inputRDD.map(lambda row: (row[1], row[0])).groupByKey().map(lambda row: sorted(list(row[1])))\
    .flatMap(lambda users: [((pair[0], pair[1]), 1) for pair in combinations(users, 2)]).reduceByKey(add)\
    .filter(lambda pair: pair[1] >= threshold).flatMap(lambda edge: [edge[0], (edge[0][1], edge[0][0])])

vertices = edgesRDD.map(lambda edge: edge[0]).distinct().zipWithIndex().map(lambda user: (str(user[1]), user[0])).toDF(['id', 'user_id']).persist(StorageLevel.DISK_ONLY)
# edges = edgesRDD.toDF(['src', 'dst'])

# (user_id, id) > user_id: id
verticesDict = dict(vertices.rdd.map(lambda row: (row[1], row[0])).collect())

# (id1, id2)
# indexing
edgesArray = edgesRDD.map(lambda edges: (verticesDict[edges[0]], verticesDict[edges[1]])).collect()
edges = sqlContext.createDataFrame(edgesArray, ["src", "dst"]).persist(StorageLevel.DISK_ONLY)

gragh = GraphFrame(vertices, edges)
# id, user_id, label
communities = gragh.labelPropagation(maxIter=5)

communitiesRDD = communities.rdd.map(lambda row: (row[2], row[1])).groupByKey().map(lambda com: sorted(list(com[1]))).sortBy(lambda com: (len(com), com))
output = '\n'.join(communitiesRDD.map(lambda com: "'" + "', '".join(com) + "'").collect())

with open(outputFilePath, 'w') as fd:
    fd.write(output)
    fd.close()

end = time.time()
print("Duration: %d seconds" % (end-start))


# cogrouped = rdd1.cogroup(rdd2)
# cogrouped.mapValues(lambda x: (list(x[0]), list(x[1]))).collect()
# ## [('foo', ([1], [4])), ('bar', ([2], [5, 6])), ('baz', ([3], []))]

File: homework4/Scala/hsinyu_chang_task1.scala
import org.apache.spark.SparkConf
import org.apache.spark.storage.StorageLevel
import org.apache.spark.sql._
import org.graphframes._
import java.io._

object hsinyu_chang_task1 {

  def main(args: Array[String]) {

    val Start = System.nanoTime

    val conf = new SparkConf()
    conf.setMaster("local[*]")
    conf.setAppName("hw")
    conf.set("spark.executor.memory", "4g")
    conf.set("spark.driver.memory", "4g")
    conf.set("spark.sql.shuffle.partitions", "4")


    val spark = SparkSession.builder.config(conf).getOrCreate()
    val sc = spark.sparkContext
    import spark.implicits._

    val threshold = args(0).toInt
    val inputFilePath = args(1)
    val outputFilePath = args(2)

    val inputFile = sc.textFile(inputFilePath)
    val header = inputFile.first()
    val inputRDD = inputFile.filter(row => row!=header).
      map(_.split(",")).map(row => (row(0), row(1))).persist(StorageLevel.DISK_ONLY)

//    build undirected graph
//    (user_id1, user_id2)
    val edgesRDD = inputRDD.map(row => (row._2, row._1)).groupByKey.map(_._2.toArray.sorted).
      flatMap(_.combinations(2).map{ case Array(p1, p2) => ((p1, p2), 1)}).reduceByKey(_+_).
      filter(_._2 >= threshold).flatMap(edge => Array(edge._1, (edge._1._2, edge._1._1)))

//    (user_id, id)
    val verticesIndex = edgesRDD.map(edge => edge._1).distinct.zipWithIndex.map(user => (user._1, user._2.toString)).persist(StorageLevel.DISK_ONLY)
//    user_id: id
    val verticesDict = verticesIndex.collect.toMap

//    id, user_id
    val vertices = verticesIndex.map(user => (user._2, user._1)).toDF("id", "user_id").persist(StorageLevel.DISK_ONLY)
//    (id1, id2)
    val edges = edgesRDD.map(edges=> (verticesDict(edges._1), verticesDict(edges._2))).toDF("src", "dst").persist(StorageLevel.DISK_ONLY)
    val graph = GraphFrame(vertices, edges)

    val communities = graph.labelPropagation.maxIter(5).run()

//    same label > same community size > order by community size
    val communitiesRDD = communities.rdd.map(row => (row(2).toString, row(1).toString)).groupByKey.
      map(com => (com._2.size, com._2.toArray.sorted.mkString("'", "', '", "'"))).groupByKey.sortBy(_._1).
      map(degree => degree._2.toArray.sorted.mkString("\n"))

    val output = communitiesRDD.collect.mkString("\n")
    val writer = new PrintWriter(new File(outputFilePath))
    writer.write(output)
    writer.close()

    val end = System.nanoTime
    println("Duration: " + ((end-Start)/1000000000.0).toInt + " seconds" )
  }
}


//    val vertices = spark.createDataFrame(List(
//      ("a", "Alice"),
//      ("b", "Bob"),
//      ("c", "Charlie"),
//      ("d", "David"),
//      ("e", "Esther"),
//      ("f", "Fanny"),
//      ("g", "Gabby")
//    )).toDF("id", "name")
//    // Edge DataFrame
//    val edges = spark.createDataFrame(List(
//      ("a", "b"),
//      ("b", "c"),
//      ("c", "b"),
//      ("f", "c"),
//      ("e", "f"),
//      ("e", "d"),
//      ("d", "a"),
//      ("a", "e")
//    )).toDF("src", "dst")

